keps/sig-scheduling/20200612-deducted-topology-manager.md

```diff
@@ -0,0 +1,177 @@
+information is necessary.
+
+# Proposal
+Kube-scheduler builtin plugin will be added to the main tree. This plugin
+implements a deducted version of TopologyManager it’s different from original topology manager algorithm.
+Plugin checks the ability to run pod only in case of single-numa-node policy on the
+node, since it is the most strict policy, it implies that the launch on the node with
+other existing policies will be successful if the condition for single-numa-node policy passed for the worker node.
+Proposed plugin will use node label to identify which topology policy is
+enabled on the node.
+To work, this plugin requires topology information of the available resource on the worker nodes.
+
+
+## Node labels
+
+Node label contains the name of the topology policy currently implemented in kubelet.
+
+Proposed Node Label may look like this:
+  `beta.kubernetes.io/topology=none|best-effort|restricted|single-numa-node`
+
+It is based on [this Proposal] (https://github.com/kubernetes/enhancements/pull/1340)
+
+To use these labels both in kube-scheduler and in kubelet string constants of these labels should be moved from pkg/kubelet/cm/topologymanager/ and pkg/kubelet/apis/config/types.go to
+pkg/apis/core/types.go
+
+## Topology format
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit).
+
+The daemon which runs outside of the kubelet will collect all necessary information on running pods, based on allocatable resources of the node and consumed resources by pods it will provide available resources in CRD, where one CRD instance represents one worker node. The name of the CRD instance is the name of the worker node.
+
+## Plugin implementation details
+
+Since topology of the node is stored in the CRD, kube-scheduler should be subscribed for updates of appropriate CRD type. Kube-scheduler will use informers which will be generated.
+
+### Topology information in NodeInfo
+
+Since not every cluster has servers with NUMA topology and not every node of the cluster with NUMA topology enables TopologyManager feature gate, topology in the NodeInfo doesn’t substitute existing representation of the resource to keep compatibility with existing scheduler plugin, but it extends NodeInfo with following field:
+
+```go
+Nodes []NUMANodeResource
+
+type NUMANodeResource struct {
+    NUMAID int
+    Resources v1.ResourceList
+}
+```
+
+
+### Description of the Algorithm
+
+The algorithm which implements single-numa-node policy is following:
+
+```go
+for _, container := range containers {
+       bitmask := bm.NewEmptyBitMask()
+       for resource, quantity := range container.Resources.Requests {
+	       resourceBitmask := bm.NewEmptyBitMask()
+	       if guarantedQoS(&container.Resources.Limits, resource, quantity) {
+		       for numaIndex, numaNodeResources := range numaMap {
```

Comment:
can we change this to match what you are proposing to add to NodeInfo (lines 100-105): a list of NUMANodeResources, and the index is NUMAID

Reply:
yes, sure, it would be better. It was intermediate map, since I tried several approaches, like prefixes with numa%d in ResourceName. Now it's not necessary, especially it may confuse here in proposal.